src/input.c

```diff
@@ -34,99 +34,188 @@
 #include "weapon.h"
 #include "utf8.h"
 
+#define KST_STRCMP_ENUM(name,str) if (!strcmp(str,#name)) return name;
+
+unsigned long hash(const char *name) {
+    unsigned long hash = 5381;
+    for (int i=0; name[i]!='\0'; i++) {
+        hash = ((hash << 5) + hash) + name[i];
+    }
+    return hash;
+}
 /* keybinding structure */
 /**
  * @brief Naev Keybinding.
  */
 typedef struct Keybind_ {
    int disabled; /**< Whether or not it's disabled. */
-   const char *name; /**< keybinding name, taken from keybind_info */
+   const char *brief, *detailed; /**< Descriptions of the keybinds */
    KeybindType type; /**< type, defined in player.h */
    SDL_Keycode key; /**< key/axis/button event number */
    SDL_Keymod mod; /**< Key modifiers (where applicable). */
 } Keybind;
 
-/* name of each keybinding */
-const char *keybind_info[][3] = {
+/**
+ * @brief Naev internal key types
+ */
+typedef enum KeySemanticType_ {
+   KST_ACCEL=0,
+   KST_LEFT,
+   KST_RIGHT,
+   KST_REVERSE,
+   KST_FACE,
+
+   KST_STEALTH,
+   KST_GAME_SPEED,
+   KST_PAUSE,
+
+   KST_AUTONAV,
+   KST_APPROACH,
+   KST_MOUSE_FLYING,
+   KST_JUMP,
+
+   KST_TARGET_NEXT,
+   KST_TARGET_PREV,
+   KST_TARGET_CLOSE,
+   KST_TARGET_SPOB,
+   KST_TARGET_JUMP,
+
+   KST_HTARGET_NEXT,
+   KST_HTARGET_PREV,
+   KST_HTARGET_CLOSE,
+
+   KST_TARGET_CLEAR,
+
+   KST_FIRE_PRIMARY,
+   KST_FIRE_SECONDARY,
+   KST_INIT_COOLDOWN,
+
+   KST_TAB_1,
+   KST_TAB_2,
+   KST_TAB_3,
+   KST_TAB_4,
+   KST_TAB_5,
+   KST_TAB_6,
+   KST_TAB_7,
+   KST_TAB_8,
+   KST_TAB_9,
+   KST_TAB_0,
+
+   KST_LOCAL_MAP,
+   KST_GLOBAL_MAP,
+
+   KST_MENU_SMALL,
+   KST_MENU_INFO,
+   KST_MENU_LUA,
+
+   KST_ESCORT_ATTACK,
+   KST_ESCORT_HALT,
+   KST_ESCORT_RETURN,
+   KST_ESCORT_CLEAR,
+
+   KST_COMM_HAIL,
+   KST_COMM_RECEIVE,
+   KST_COMM_UP,
+   KST_COMM_DOWN
+
+   KST_ZOOM_IN,
+   KST_ZOOM_OUT,
+
+   KST_FULLSCREEN,
+
+   KST_SCREENSHOT,
+
+   KST_PASTE,
+
+} KeySemanticType;
+
+
+/* Description of each key semantic type */
+const char *keybind_info[KST_PASTE+1][2] = {
    /* Movement */
-   { "accel", N_("Accelerate"), N_("Makes your ship accelerate forward.") },
-   { "left", N_("Turn Left"), N_("Makes your ship turn left.") },
-   { "right", N_("Turn Right"), N_("Makes your ship turn right.") },
-   { "reverse", N_("Reverse"), N_("Makes your ship face the direction you're moving from. Useful for braking.") },
-   { "stealth", N_("Stealth"), N_("Tries to enter stealth mode.") },
+   { N_("Accelerate"), N_("Makes your ship accelerate forward.") },
```

Comment:
It would probably be better to explicitly include the KeySemanticType here to avoid bugs, although you would have to change it to a struct instead of an array of chars. You could use `#if DEBUGGING` guards to also make sure the order is correct on loading (so that the position of the keybind with KST_PASTE matches the enum value of KST_PASTE).

Reply:
That wouldn't be needed; you could initialize it with [KST_X]={appropriate}, and I will do that in a bit.